[Free Trial] Add Free trial store creation flow tracks

ThomazFB opened the pull request:
Fix issue #9068 

iOS counterpart of https://github.com/woocommerce/woocommerce-android/pull/8648

⚠️ I used the `issue/9244-update-site-creation-flow` as the targeted branch to enable testing the current Free Trial creation flow with the actual Analytics changes. Because of that, this PR should only be merged once https://github.com/woocommerce/woocommerce-ios/pull/9317 is finished.

Why
==========
We need to easily filter the paid store creation flow analytics from the free trial one when looping back to see the store creation.


How
==========
Adds a `isFreeTrial` parameter to the following four store creation track events:

- `login_prologue_create_site_tapped`
- `login_woocommerce_site_created`
-  `site_creation_failed`
-  `site_creation_dismissed`

With the changes introduced here, the `isFreeTrial` will be true when the Free Trial Feature Flag is active and uses the `native` creation flow, but when we remove the flag, we should replace the parameter with always true for all the native flow. In case we come back to providing two different store creation flows back again, then this parameter shall be set depending on the creation type.

How to Test
==========
1. Create a Free Trial store and verify if the Track events keep exactly the same as before, but the described four track events contain the `isFreeTrial` parameter value as `true`.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Adjust WooAnalyticsEvent+StoreCreation to expect the free trial track argument

ThomazFB committed:
Update StoreCreationCoordinator with Free Trial creation checks

ThomazFB committed:
Adjust WooAnalyticsEvent+StoreCreation to expect the free trial creation start track argument

ThomazFB committed:
Update AuthenticationManager with Free Trial creation checks

ThomazFB committed:
Fix hound issues

Ecarrion commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -414,13 +416,19 @@ private extension StoreCreationCoordinator {
+                analytics.track(event: .StoreCreation.siteCreationFailed(source: source.analyticsValue,
+                                                                         error: error,
+                                                                         flow: .native,
+                                                                         isFreeTrial: isFreeTrialCreation))
```
I think this will always be true as it is inside the `createFreeTrialStore`

Ecarrion commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -414,13 +416,19 @@ private extension StoreCreationCoordinator {
+            analytics.track(event: .StoreCreation.siteCreationFailed(source: source.analyticsValue,
+                                                                     error: error,
+                                                                     flow: .native,
+                                                                     isFreeTrial: isFreeTrialCreation))
```
Same as this one.

Ecarrion commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -472,7 +480,11 @@ private extension StoreCreationCoordinator {
+            analytics.track(event: .StoreCreation.siteCreationFailed(source: source.analyticsValue,
+                                                                     error: error,
+                                                                     flow: .native,
+                                                                     isFreeTrial: isFreeTrialCreation))
```
And I think this will always be `false`

Ecarrion commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -679,7 +691,8 @@ private extension StoreCreationCoordinator {
     func showSuccessView(from navigationController: UINavigationController, site: Site) {
-        analytics.track(event: .StoreCreation.siteCreated(source: source.analyticsValue, siteURL: site.url, flow: .native))
+        let isFreeTrialCreation = featureFlagService.isFeatureFlagEnabled(.freeTrial)
+        analytics.track(event: .StoreCreation.siteCreated(source: source.analyticsValue, siteURL: site.url, flow: .native, isFreeTrial: isFreeTrialCreation))
```
Also `always` false.

Ecarrion reviewed (commented):
Thanks for the PR @ThomazFB, I have some comments!

----

Do you think it is worth it to extract 

```
let isFreeTrialCreation = featureFlagService.isFeatureFlagEnabled(.freeTrial)
```

Into an instance variable and use that instead of fetching the status in every place? I think it will make the code easier to read and maintain.

----

I think you need to merge trunk's latest changes, as there are a few additions on `waitForSiteToBecomeJetpackSite` that involves tracking events.

ThomazFB committed:
Simplify free trial feature flag check

ThomazFB committed:
Remove boolean redundancies from free trial value check

ThomazFB committed:
Merge branch 'trunk' into issue/9068-free-trial-store-creation-tracks

ThomazFB committed:
Cover waitForSiteToBecomeJetpackSite free trial analytics scenario

ThomazFB committed:
Fix hound issues

ThomazFB commented:
Thanks @Ecarrion, for your careful review. Sorry for the redundant booleans. I went adding every parameter usage with the 
feature flag checks without verifying the context, which makes the check unnecessary 🤦 .

I addressed all your comments. The PR is updated and ready for another round!

Ecarrion reviewed (approved):
Looks good @ThomazFB! Thanks for the help!
